File: go/trafficlight/trafficlight/machine.go
package trafficlight

import (
	"time"

	"github.com/frame-lang/frame-demos/go/web/traffic/framelang"
)

const (
	begin framelang.FrameState = iota
	red
	green
	yellow
	flashingRed
	working
)

type TrafficLight interface {
	Start(mom *MOM)
	Timer()
	SystemError()
	SystemRestart()
}

type actions interface {
	enterRed()
	enterGreen()
	enterYellow()
	enterFlashingRed()
	exitFlashingRed()
	startWorkingTimer()
	stopWorkingTimer()
	startFlashingTimer()
	stopFlashingTimer()
	changeColor(color string)
	startFlashing()
	stopFlashing()
	changeFlashingAnimation()
	log(msg string)
}

type trafficLightStruct struct {
	_state_    framelang.FrameState
	mom        *MOM
	ticker     *time.Ticker
	flashColor string
}

func New() TrafficLight {
	m := new(trafficLightStruct)
	// Verify TrafficLightStruct implements actions interface
	var _ actions = m
	m.mom = nil
	m.ticker = nil
	m.flashColor = ""
	return m
}

//===================== Interface Block ===================//

func (m *trafficLightStruct) Start(mom *MOM) {
	params := make(map[string]interface{})
	params["mom"] = mom
	e := framelang.FrameEvent{Msg: ">>", Params: params}
	m._mux_(&e)
}

func (m *trafficLightStruct) Timer() {
	e := framelang.FrameEvent{Msg: "timer"}
	m._mux_(&e)
}

func (m *trafficLightStruct) SystemError() {
	e := framelang.FrameEvent{Msg: "systemError"}
	m._mux_(&e)
}

func (m *trafficLightStruct) SystemRestart() {
	e := framelang.FrameEvent{Msg: "systemRestart"}
	m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *trafficLightStruct) _mux_(e *framelang.FrameEvent) {
	switch m._state_ {
	case begin:
		m._begin_(e)
	case red:
		m._red_(e)
	case green:
		m._green_(e)
	case yellow:
		m._yellow_(e)
	case flashingRed:
		m._flashingRed_(e)
	case working:
		m._working_(e)
	}
}

//===================== Machine Block ===================//

func (m *trafficLightStruct) _begin_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">>":
		m.mom = e.Params["mom"].(*MOM)
		m.startWorkingTimer()
		m._transition_(red)
		return
	}
}

func (m *trafficLightStruct) _red_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.enterRed()
		return
	case "timer":
		m._transition_(green)
		return
	}
	m._working_(e)

}

func (m *trafficLightStruct) _green_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.enterGreen()
		return
	case "timer":
		m._transition_(yellow)
		return
	}
	m._working_(e)

}

func (m *trafficLightStruct) _yellow_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.enterYellow()
		return
	case "timer":
		m._transition_(red)
		return
	}
	m._working_(e)

}

func (m *trafficLightStruct) _flashingRed_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.enterFlashingRed()
		m.stopWorkingTimer()
		m.startFlashingTimer()
		return
	case "<":
		m.exitFlashingRed()
		m.stopFlashingTimer()
		m.startWorkingTimer()
		return
	case "timer":
		m.changeFlashingAnimation()
		return
	case "systemRestart":
		m._transition_(red)
		return
	}
}

func (m *trafficLightStruct) _working_(e *framelang.FrameEvent) {
	switch e.Msg {
	case "systemError":
		m._transition_(flashingRed)
		return
	}
}

//=============== Machinery and Mechanisms ==============//

func (m *trafficLightStruct) _transition_(newState framelang.FrameState) {
	m._mux_(&framelang.FrameEvent{Msg: "<"})
	m._state_ = newState
	m._mux_(&framelang.FrameEvent{Msg: ">"})
}

/********************
// Sample Actions Implementation
package trafficlight

func (m *trafficLightStruct) enterRed()  {}
func (m *trafficLightStruct) enterGreen()  {}
func (m *trafficLightStruct) enterYellow()  {}
func (m *trafficLightStruct) enterFlashingRed()  {}
func (m *trafficLightStruct) exitFlashingRed()  {}
func (m *trafficLightStruct) startWorkingTimer()  {}
func (m *trafficLightStruct) stopWorkingTimer()  {}
func (m *trafficLightStruct) startFlashingTimer()  {}
func (m *trafficLightStruct) stopFlashingTimer()  {}
func (m *trafficLightStruct) changeColor(color string)  {}
func (m *trafficLightStruct) startFlashing()  {}
func (m *trafficLightStruct) stopFlashing()  {}
func (m *trafficLightStruct) changeFlashingAnimation()  {}
func (m *trafficLightStruct) log(msg string)  {}
********************/

File: javascript/trafficlight/trafficlight/machine.js
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

const FrameEvent = require("../framelang/framelang");

class TrafficLight {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    #flashColor;
    #mom;
    #ticker;
    
    constructor () {
        // Create and intialize start state compartment.
        
        this.#state = this.#sBegin_;
        this.#compartment = new TrafficLightCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.#flashColor = "";
        this.#mom = null;
        this.#ticker = null;
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    start(mom) {
        let e = FrameEvent(">>",{"mom":mom});
        this.#mux(e);
    }
    
    stop() {
        let e = FrameEvent("stop",null);
        this.#mux(e);
    }
    
    tick() {
        let e = FrameEvent("tick",null);
        this.#mux(e);
    }
    
    systemError() {
        let e = FrameEvent("systemError",null);
        this.#mux(e);
    }
    
    systemRestart() {
        let e = FrameEvent("systemRestart",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sBegin_:
                this.#sBegin_(e);
                break;
            case this.#sRed_:
                this.#sRed_(e);
                break;
            case this.#sGreen_:
                this.#sGreen_(e);
                break;
            case this.#sYellow_:
                this.#sYellow_(e);
                break;
            case this.#sFlashingRed_:
                this.#sFlashingRed_(e);
                break;
            case this.#sEnd_:
                this.#sEnd_(e);
                break;
            case this.#sWorking_:
                this.#sWorking_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sBegin_(e) {
        switch (e._message) {
            case ">>":
                {
                this.startWorkingTimer_do();
                let compartment =  new TrafficLightCompartment(this.#sRed_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sRed_(e) {
        switch (e._message) {
            case ">":
                {
                this.enterRed_do();
                return;
                }
                
            case "tick":
                {
                let compartment =  new TrafficLightCompartment(this.#sGreen_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
        this.#sWorking_(e);
        
    }
    
    #sGreen_(e) {
        switch (e._message) {
            case ">":
                {
                this.enterGreen_do();
                return;
                }
                
            case "tick":
                {
                let compartment =  new TrafficLightCompartment(this.#sYellow_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
        this.#sWorking_(e);
        
    }
    
    #sYellow_(e) {
        switch (e._message) {
            case ">":
                {
                this.enterYellow_do();
                return;
                }
                
            case "tick":
                {
                let compartment =  new TrafficLightCompartment(this.#sRed_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
        this.#sWorking_(e);
        
    }
    
    #sFlashingRed_(e) {
        switch (e._message) {
            case ">":
                {
                this.enterFlashingRed_do();
                this.stopWorkingTimer_do();
                this.startFlashingTimer_do();
                return;
                }
                
            case "<":
                {
                this.exitFlashingRed_do();
                this.stopFlashingTimer_do();
                this.startWorkingTimer_do();
                return;
                }
                
            case "tick":
                {
                this.changeFlashingAnimation_do();
                return;
                }
                
            case "systemRestart":
                {
                let compartment =  new TrafficLightCompartment(this.#sRed_);
                
                
                this.#transition(compartment);
                return;
                }
                
            case "stop":
                {
                let compartment =  new TrafficLightCompartment(this.#sEnd_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sEnd_(e) {
        switch (e._message) {
            case ">":
                {
                this.stopWorkingTimer_do();
                return;
                }
                
        }
    }
    
    #sWorking_(e) {
        switch (e._message) {
            case "stop":
                {
                let compartment =  new TrafficLightCompartment(this.#sEnd_);
                
                
                this.#transition(compartment);
                return;
                }
                
            case "systemError":
                {
                let compartment =  new TrafficLightCompartment(this.#sFlashingRed_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    enterRed_do() { throw new Error('Action not implemented.'); }
    enterGreen_do() { throw new Error('Action not implemented.'); }
    enterYellow_do() { throw new Error('Action not implemented.'); }
    enterFlashingRed_do() { throw new Error('Action not implemented.'); }
    exitFlashingRed_do() { throw new Error('Action not implemented.'); }
    startWorkingTimer_do() { throw new Error('Action not implemented.'); }
    stopWorkingTimer_do() { throw new Error('Action not implemented.'); }
    startFlashingTimer_do() { throw new Error('Action not implemented.'); }
    stopFlashingTimer_do() { throw new Error('Action not implemented.'); }
    changeColor_do(color) { throw new Error('Action not implemented.'); }
    startFlashing_do() { throw new Error('Action not implemented.'); }
    stopFlashing_do() { throw new Error('Action not implemented.'); }
    changeFlashingAnimation_do() { throw new Error('Action not implemented.'); }
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};

//=============== Compartment ==============//

class TrafficLightCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}




class TrafficLightController extends TrafficLight {

	constructor() {
	  super()
	}
	enterRed_do() {
        console.log("enterRed_do()")
    }
	enterGreen_do() {
        console.log("enterGreen_do()")
    }
	enterYellow_do() {
        console.log("enterYellow_do()")
    }
	enterFlashingRed_do() {}
	exitFlashingRed_do() {}
	startWorkingTimer_do() {
        this.ticker = setInterval(() => {
            this.tick()
        }, 1000)
    }
	stopWorkingTimer_do() {
        clearInterval(this.ticker)
    }
	startFlashingTimer_do() {}
	stopFlashingTimer_do() {}
	changeColor_do(color) {}
	startFlashing_do() {}
	stopFlashing_do() {}
	changeFlashingAnimation_do() {}
	log_do(msg) {}
};

module.exports = TrafficLightController


